engine/occurrence.ftl

```diff
@@ -203,13 +203,54 @@
     [#return occurrence]
 [/#function]
 
+[#function extendOccurrenceAttributes provider attributes=[] extensions=[]]
+    [#local result = []]
+    [#if extensions?has_content]
+        [#local providerExtensions = extensions[provider]![]]
+        [#list attributes as attribute]
+
+            [#local attributeExtension = 
+                providerExtensions
+                ?filter(e -> asArray(attribute.Names)
+                    ?seq_contains(e.Names))]
+
+            [#if attributeExtension?has_content]
```

Comment:
Will this support nested attributes? I.e. Children attributes?

Reply:
You're right, it wont.

I might move the Extensions on to each attribute directly, as it will allow me to re-use the recursion already in place for this purpose on the  function `getCompositeObject()`

Example:

<details>
<summary>The Component definition currently constructed</summary>
<br>

```json
{
  "ResourceGroups" : {
    "DEFAULT_RESOURCE_GROUP" : {
      "Attributes" : [
       ...
      ],
      "Extensions" : [
        "azure" :  {
          ... azureextensions ...
        }
      ]
    }
  }
}
```

</details>

<details>
<summary>Extensions assigned to each attribute</summary>
<br>

```json
{
  "ResourceGroups" : {
    "DEFAULT_RESOURCE_GROUP" : {
      "Attributes" : [
        "engine" {
          "Names" : [ "Engine" ],
          "Type" : "string",
          "Values" : [ "engine1", "engine2" ],
          "Extensions" : [
            {
               "azure" : [ "engine3", "engine4" ]
            }
          ]
        }
      ]
    }
  }
}
```

</details>

And then I could move the current extension value logic into `getCompositeObject()` to perform the merge of values.

<details>
<summary>The final composite object</summary>
<br>

```json
{
  "ResourceGroups" : {
    "DEFAULT_RESOURCE_GROUP" : {
      "Attributes" : [
        "engine" {
          "Names" : [ "Engine" ],
          "Type" : "string",
          "Values" : [ "engine1", "engine2", "azure:engine3", "azure:engine4" ],
          "Extensions" : [
            {
               "azure" : [ "engine3", "engine4" ]
            }
          ]
        }
      ]
    }
  }
}
```

</details>

If I keep the "Extensions" at the same level of "Attributes", this will require updating `getCompositeObject()` too much. As the recursion drills down on the child attributes, it would likewise need to drill down on the child extensions object, which would mean new function parameters.

@ml019 @roleyfoley any concerns with this?

Reply:
getCompositeObject isn't the right place to do this. It expects a list of attributes full stop.

Any nested attribute values need to be added before the call to getCompositeObject.

I'd suggest putting the shared extensions under the provider specific section of componentConfiguration - they logically belong with the provider not the shared implementation.

Reply:
I have reworked this to allow for for child attributes & have tested it as follows:

<details>
<summary>Testing Extending the Value Sets for the DB Component</summary>
<br>

```
[@addResourceGroupAttributeValues
    type=DB_COMPONENT_TYPE
    provider=AZURE_PROVIDER
    extensions=[
        {
            "Names" : "Backup",
            "Children" : [
                {
                    "Names" : "DeletionPolicy",
                    "Values" : [ "notadeletionpolicy", "notreal"]
                }
            ]            
        },
        {
            "Names" : "Cluster",
            "Children" : [
                {
                    "Names" : "ScalingPolicies",
                    "Children" : [
                        {
                            "Names" : "Type",
                            "Values" : [ "Slippery-Dip" ]
                        }
                    ]
                }
            ]
        }
    ]
/]
```

</details>


<details>
<summary>Results for Backup.DeletionPolicy (direct children)</summary>
<br>

```
       {
          "Names": "Backup",
          "Children": [
            ...
            {
              "Names": "DeletionPolicy",
              "Type": "string",
              "Values": [
                "Snapshot",
                "Delete",
                "Retain",
                "azure:notadeletionpolicy",
                "azure:notreal"
              ],
              "Default": "Snapshot"
            }
            ...
          ]
        }
```

</details>

<details>
<summary>Results for Cluster.ScalingPolicies.Type (SubObjects = true) with </summary>
<br>

```
        {
          "Names": "Cluster",
          "Description": "Cluster specific configuration when using a clustered database engine",
          "Children": [
            {
              "Names": "ScalingPolicies",
              "Subobjects": true,
              "Children": [
                {
                  "Names": "Type",
                  "Type": "string",
                  "Values": [
                    "Stepped",
                    "Tracked",
                    "Scheduled",
                    "azure:Slippery-Dip"
                  ],
                  "Default": "Stepped"
                }
               ...
```

</details>

Reply:
also, I've updated the extension function to simply `extendAttributes()` as it should also work for Reference Data, although this was out of scope and is untested.

Reply:
Good thought re reference data - so we have a likely use case for that?

Reply:
@ml019 not quite, but I think we could actually do a fair bit with the idea. 

For example, all of the ARRAY_OF_OBJECT_TYPE attributes are Reference Data. For things like the "SKUProfile" (the combination of hardware make/model/unit used by many Azure resources, but also common hardware term in general) I've created in the Azure Provider - its not a unique-to-azure concept, its just not used by the AWS provider. Instead of having such things in the plugins we could flesh out their definition (much like we would with a new component type) so its an understood, and then expect providers that use it to extend its value sets with arrays of object types. As it stands there is little reason to do this however, but in a world with more plugins I think this could help keep definitions and terms at a "solution" and common-ground conceptual level.

The other use-case I had thought of was removing things like "Regions" from masterdata, and having them implemented as ARRAY_OF_OBJECT_TYPE by each provider. Both approaches have their pros/cons but I know we've discussed getting rid of masterdata down the road.